photon.c

```diff
@@ -0,0 +1,135 @@
+#include <stdlib.h>
+#include <stdio.h>
+#include <unistd.h>
+#include <sys/socket.h>
+#include <sys/un.h>
+#include <inttypes.h>
+
+#include "utarray.h"
+#include "common.h"
+#include "state/task_queue.h"
+#include "state/db.h"
+#include "event_loop.h"
+#include "task.h"
+#include "photon.h"
+#include "io.h"
+
+typedef struct {
+  event_loop *loop;
+  db_handle *db;
+  UT_array *task_queue;
+} local_scheduler_state;
+
+void init_local_scheduler(local_scheduler_state *s) {
+  s->loop = event_loop_create();
+};
+
+void process_message(event_loop *loop, int client_sock, void *context, int events) {
+  local_scheduler_state *s = context;
+
+  uint8_t *message;
+  int64_t type;
+  int64_t length;
+  read_message(client_sock, &type, &length, &message);
+
+  switch (type) {
+    case SUBMIT_TASK: {
+        task_spec *task = (task_spec *) message;
+        CHECK(task_size(task) == length);
+        unique_id id = globally_unique_id();
+        task_queue_submit_task(s->db, id, task);
+      }
+      break;
+    case TASK_DONE: {
+      }
+      break;
+    case DISCONNECT_CLIENT: {
+        LOG_INFO("Disconnecting client on fd %d", client_sock);
+        event_loop_remove_file(loop, client_sock);
+      }
+      break;
+    case LOG_MESSAGE: {
+      }
+      break;
+    default:
+      /* This code should be unreachable. */
+      CHECK(0);
+  }
+  free(message);
+}
+
+void new_client_connection(event_loop *loop, int listener_sock, void *context, int events) {
+  local_scheduler_state *s = context;
+  int new_socket = accept(listener_sock, NULL, NULL);
+  if (new_socket < 0) {
+    if (errno != EWOULDBLOCK) {
+      LOG_ERR("accept failed");
+      exit(-1);
+    }
+    return;
+  }
+  event_loop_add_file(s->loop, new_socket, EVENT_LOOP_READ, process_message, s);
+  LOG_INFO("new connection with fd %d", new_socket);
+}
+
+void start_server(const char* socket_name, const char* redis_addr, int redis_port) {
+  int fd = socket(AF_UNIX, SOCK_STREAM, 0);
+  if (fd == -1) {
+    LOG_ERR("socket error");
+    exit(-1);
+  }
+  int on = 1;
+  if (setsockopt(fd, SOL_SOCKET, SO_REUSEADDR, (char*) &on, sizeof(on)) < 0) {
+    LOG_ERR("setsockopt failed");
+    close(fd);
+    exit(-1);
+  }
+  struct sockaddr_un addr;
+  memset(&addr, 0, sizeof(addr));
+  addr.sun_family = AF_UNIX;
+  strncpy(addr.sun_path, socket_name, sizeof(addr.sun_path) - 1);
+  unlink(socket_name);
+  bind(fd, (struct sockaddr*) &addr, sizeof(addr));
+  listen(fd, 5);
```

Comment:
There is a function called `bind_ipc_sock` in `io.h` that you can use here, if you want.